core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -383,11 +384,44 @@ class TransactionCoordinator(brokerId: Int,
 
   def partitionFor(transactionalId: String): Int = txnManager.partitionFor(transactionalId)
 
+  private def expireTransactions(): Unit = {
+
+    txnManager.transactionsToExpire().foreach{ idAndMetadata =>
+      idAndMetadata.metadata synchronized {
+        if (!txnManager.isCoordinatorLoadingInProgress(idAndMetadata.transactionalId)
+          && idAndMetadata.metadata.pendingState.isEmpty) {
+          idAndMetadata.metadata.producerEpoch = (idAndMetadata.metadata.producerEpoch + 1).toShort
+          idAndMetadata.metadata.prepareTransitionTo(Ongoing)
+          txnManager.appendTransactionToLog(idAndMetadata.transactionalId, idAndMetadata.metadata, (errors: Errors) => {
+            if (errors != Errors.NONE)
+              warn(s"failed to append transactionalId ${idAndMetadata.transactionalId} to log during transaction expiry. errors:$errors")
+            else
+              handleEndTransaction(idAndMetadata.transactionalId,
+                idAndMetadata.metadata.pid,
+                idAndMetadata.metadata.producerEpoch,
+                TransactionResult.ABORT,
+                (errors: Errors) => {
+                  if (errors != Errors.NONE)
+                    warn(s"rollback of transactionalId: ${idAndMetadata.transactionalId} failed during transaction expiry. errors: $errors")
+                }
+              )
+          })
+        }
+      }
+    }
+  }
+
   /**
    * Startup logic executed at the same time when the server starts up.
    */
   def startup(enablePidExpiration: Boolean = true) {
     info("Starting up.")
+    scheduler.startup()
```

Comment:
I see.. since we area already on-hold on the metadata object here, maybe we do not need to directly call the general `handleEndTxn` but we can extract part of that logic in which we have already read out the metadata object from the cache, and then just call that from here. In that case, we can still move the scheduler / etc stuff inner?